README
(emacs users can use -*- outline -*- mode for this file)

$Id: README,v 1.7 2002/01/07 20:16:49 erikm Exp $

* Info
======

** What is blob?
----------------

Blob is the Boot Loader OBject, the boot loader for the LART. Blob is
able to boot a Linux kernel stored in flash or RAM and provide that
kernel with a ramdisk (again from flash or RAM).

Blob is copyrighted by Jan-Derk Bakker and Erik Mouw. Blob is released
with a slightly modified GNU GPL license: we don't consider the
operating systems that blob boots as a derived work. Later on, several
other people also contributed to blob.

Blob started its life as a boot loader for the LART, but nowadays it
has been ported to the Intel Assabet SA-1110 evaluation platform, the
Intel Brutus SA-1100 evaluation platform, the PLEB board, the Nesa
board, the TuxScreen (aka Shannon), and to the CreditLART board.




** Where is the latest blob source available?
---------------------------------------------

The latest and greatest blob source is available from SourceForge, see
http://www.sourceforge.net/projects/blob/ . The latest source is
available from anonymous CVS. First log in to the CVS server:

  cvs -d:pserver:anonymous@cvs.blob.sourceforge.net:/cvsroot/blob login

There is no password, so just press enter. Now check out the blob source:

  cvs -z3 -d:pserver:anonymous@cvs.blob.sourceforge.net:/cvsroot/blob co blob

If you're using the blob CVS source, it's a good idea to subscribe to
the blob-cvs-commit mailing list so you know about blob patches. See
http://lists.sourceforge.net/lists/listinfo/blob-cvs-commit . The
general blob discussion is done on the LART mailing list, see
http://www.lart.tudelft.nl/list/ for more information.

There is also a blob IRC channel: log on to irc.openprojects.net and
join #blob. Note that this is a strictly on-topic development IRC
channel, not a general blob help channel.

Blob even has a home page: http://www.lart.tudelft.nl/lartware/blob/ .




** So what is LART?
-------------------

LART is the Linux Advanced Radio Terminal, a small low power computing
element used in the MMC project and the Ubiquitous Communications
programme (see http://www.lart.tudelft.nl, http://www.mmc.tudelft.nl,
and http://www.ubicom.tudelft.nl ).

LART features:

- 10x7.5 cm (that's 4x3 inch in Stonehenge Units)
- 220 MHz Digital StrongARM SA-1100 CPU
- 4 Mbyte flash memory
- 32 Mbyte DRAM memory
- Low power: peak power consumption is 1W




* Building blob
===============

** Prerequisites
----------------

- A native ARM/StrongARM Linux system with gcc 2.95.2, and
  binutils 2.9.5.0.22 or better
- Or any UNIX system with cross compiling binutils 2.9.5.0.22 (or
  better) and gcc 2.95.2 installed
- GNU make (although some vendor supplied make utilities will do)
- GNU autoconf and automake (if you build blob from CVS)
- A configured Linux kernel source tree: the latest linux-2.4.* kernel
  will usually do. If not, apply the appropriate -rmk patch (see the
  LART website for details). Linux-2.2.* kernels will NOT work, but
  it's considered obsolete for arm-linux anyway.

We tested blob with a native system (Corel/Rebel Netwinder) and with
several i386-linux to arm-linux cross compiler systems, but the GNU
tools are that good that we don't think that a sun-sparc-solaris to
arm-linux cross compiler will fail.




** Generating configure and Makefiles
-------------------------------------

This step is only necessary if you build blob from CVS.

- Run "tools/rebuild-gcc" ttwwiiccee




** Configuring and compiling the package
----------------------------------------

With a cross compiler tool chain (using tcsh as shell):

- setenv CC /path/to/cross/gcc
- setenv OBJCOPY /path/to/cross/objcopy
- Run "./configure --with-linux-prefix=/path/to/armlinux/source \
       --with-board=boardname arm-unknown-linux-gnu"

  There are currently a couple of valid board names, choose from:
  assabet, brutus, creditlart, lart, nesa, pleb, or shannon. If the
  board name is ommited, lart will be chosen as a default.

  If you want to do some serious hacking on blob, consider using the
  "--enable-maintainer-mode" flag. This will automatically regenerate
  Makefiles and configure scripts if you change the source files. You
  need autoconf (>= 2.13) and automake (>= 1.4) for this feature.

  Note that the linux kernel needs to be configured or otherwise
  blob will not compile. To get a configured kernel tree, unpack the
  kernel and run:

  make ARCH=arm CROSS_COMPILE=arm-linux- mrproper
  make ARCH=arm CROSS_COMPILE=arm-linux- lart_config
  yes no | make ARCH=arm CROSS_COMPILE=arm-linux- oldconfig

- Run "make"

If you use the bash shell, use "export FOO=bar" instead of
"setenv FOO bar".

With a native ARM Linux tool chain:

- Run "./configure --with-board=boardname" 
- Run "make"

The binary image is in src/blob; src/blob-start-elf32 and
src/blob-rest-elf32 are the two parts of the images with complete ELF
headers. To disassemble "blob-start-elf32", use:

  arm-linux-objdump -D -S blob-start-elf32

To see the actual hex codes of blob, use:

  od -t x4 blob




** Installing
-------------

*** LART
--------

The current wisdom to install blob on a LART is:

- Connect the JTAG dongle to the LART
- Connect the other end of the JTAG dongle to the parallel port of
  your PC
- Power up the LART
- Use the jflash utility (available from the LART web site) to write
  blob (you usually need  to be root for this): jflash blob

The JTAG flash burn code however is now worked out as a set of Linux
executables provided by the JTAG flash project located at the LART page
as well as JTAG executables ported to support the TuxScreen screen phone.

The LART project initially used the following wisdom to install blob:

Required hardware & software:

- The LART itself with 4 Mbyte flash memory
- An external 128 kbyte flash board
- A PCI 7200 (???) digital I/O card with a Linux driver
- A flash burn program for this I/O card

The external flash board is connected to the PCI 7200 card and blob is
written into the flash memory using the flash burn program. The
external flash board is connected to the LART low speed interface. The
external flash chip is mapped at address 0x00000000, and the internal
flash is re-mapped at 0x08000000. As soon as the LART boots, the
external flash is copied to the first 128 kbyte of the internal
flash. The next time the LART is started without external flash board,
it starts from its internal flash which now contains the just
downloaded blob.

Why this strange way to download blob? We first tried to use the
SA-1100 JTAG interface to program the flash directly, but soon found
out that it would take weeks to write a decent JTAG tool because JTAG
is a real brain-damaged protocol (it was designed by a committee, need
we say more?). To meet a deadline, we decided to make a special board
with 128 kbyte external flash memory (and an LCD interface).


*** Assabet
-----------

(From Justin Seger:) The best way is to use the JTAG cable:
- Connect the JTAG cable from the Assabet to your hosts parallel port
- Power up the Assabet
- Use the jflash utility to write blob: jflash-linux blob
- Power cycle the Assabet; you should see the the bootloader starting
  up with the output on the serial port.


*** SHANNON (TuxScreen web phone)
-----------

http://TuxScreen.net/

The Shannon comes with Inferno (http://www.vitanuova.com/inferno/)
installed on it. You can DL and install a hosted version of Inferno
and then use the sboot remote interface to install blob for the first
time. Afterwards blob can reinstall itself.

http://TuxScreen.net/wiki/view//InfernoRemote

Alternately, you can use JTAG hardware to do the install if you have
this equipment.

http://TuxScreen.net/wiki/view/Main/JTAG

** Making a distribution
------------------------

This is only needed when you want to make a tar file from the current
blob sources.

- First configure the package
- Run "make dist"




* Using blob
============

** Booting
----------

First connect a terminal (or a terminal emulator like miniterm or
Seyon) to the serial port. Use the following settings for your
terminal: 9600 baud, 8 data bits, no parity, 1 stop bit, no start bits
(9600 8N1, a pretty standard setting for Unix systems). If possible,
use VT100 terminal emulation. Switch on the power to the SA-11x0
board. The board should respond with:

  Consider yourself LARTed!

  blob version 2.0.3
  Copyright (C) 1999 2000 2001 Jan-Derk Bakker and Erik Mouw
  Copyright (C) 2000 Johan Pouwelse
  blob comes with ABSOLUTELY NO WARRANTY; read the GNU GPL for details.
  This is free software, and you are welcome to redistribute it
  under certain conditions; read the GNU GPL for details.
  Memory Map:
    0x08000000 @ 0xC0000000 (8MB)
    0x08000000 @ 0xC1000000 (8MB)
    0x08000000 @ 0xC8000000 (8MB)
    0x08000000 @ 0xC9000000 (8MB)
  Loading blob from flash . done
  Loading kernel from flash ....... done
  Loading ramdisk from flash ............... done
  Autoboot in progress, press any key to stop ...

If you don't press a key within 10 seconds, blob will automatically
start the Linux kernel:

  Starting kernel ...

  Uncompressing Linux...done.
  Now booting the kernel
  ...

However, if you press the <enter> key, you will get the blob
prompt:

  Autoboot aborted
  Type "help" to get a list of commands
  blob> 




** Commands
-----------

Blob knows several commands, typing "help" (without the ") will show
you which:

  Help for blob 2.0.3, the LART bootloader
  The following commands are supported:
  * boot [kernel options]           Boot Linux with optional kernel options
  * clock PPCR MDCNFG MDCAS0 MDCAS1 MDCAS2
                                    Set the SA1100 core clock and DRAM timings
                                    (WARNING: dangerous command!)
  * download {blob|kernel|ramdisk}  Download blob/kernel/ramdisk image to RAM
  * flash {kernel|ramdisk}          Copy blob/kernel/ramdisk from RAM to flash
  * help                            Get this help
  * reblob                          Restart blob from RAM
  * reboot                          Reboot system
  * reload {blobkernel|ramdisk}     Reload blob/kernel/ramdisk from flash to RAM
  * reset                           Reset terminal
  * speed                           Set download speed
  * status                          Display current status


*** "boot"
----------

Boot the Linux kernel. You can supply extra parameters to the Linux
kernel; if you don't, the kernel will use it's default command
line. Blob will respond with:

  blob> boot

  Starting kernel ...

  Uncompressing Linux...done.
  Now booting the kernel
  ...


*** "clock"
-----------

This an experimental command to set the SA1100 core clock and DRAM
timings. We've used it to test clock scaling. This command writes the
exact values supplied on the command line to the PPCR, MDCNFG, MDCAS0,
MDCAS1, and MDCAS2 registers, but it doesn't check the validity of the
values. Example (that will crash your system for sure):

  blob> clock 0x11111111 0x22222222 0x33333333 0x44444444 0x55555555

WARNING: This command is DANGEROUS and HIGHLY EXPERIMENTAL. Don't use
it unless you have a VERY thorough understanding on the inner workings
of the SA1100 CPU! It works for us, YMMV. If it breaks your CPU, don't
say that we didn't warn you!


*** "download"
--------------

Download a uuencoded blob, kernel, or ramdisk to RAM. This command
needs an extra parameter: "blob", "kernel", or "ramdisk". Blob will
respond with:

  blob> download kernel
  Switching to 115200 baud
  You have 60 seconds to switch your terminal emulator to the same speed and
  start downloading. After that blob will switch back to 9600 baud.

Switch your terminal emulator to the indicated speed and start
downloading the kernel or ramdisk. With minicom, you can use the ASCII
download method, or use another shell to download the file:

  uuencode zImage zImage > /dev/ttyS1

Of course, use the correct serial port. If the download is successful,
blob will respond with:

  (Please switch your terminal emulator back to 9600 baud)
  Received 65536 (0x00010000) bytes.

If an error occurs during downloading, blob will respond with:

  (Please switch your terminal emulator back to 9600 baud)
  *** Uudecode receive failed

A failed download session can have several reasons: the file is too
big, the download speed too high (see the "speed" command), or the
uuencoded file to be downloaded is corrupt. Correct the error and
retry.

A downloaded kernel or ramdisk can be written to flash with the
"flash" command, or it can directly be used to boot with the "boot"
command.


*** "flash"
-----------

Write blob, kernel, or ramdisk from RAM to flash memory. This command
needs an extra parameter: "blob", "kernel" or "ramdisk". Blob will
respond with:

  blob> flash kernel
  Saving kernel to flash ..... .... done

This won't work on all architectures, check the RELEASE-NOTES.


*** "reblob"
------------

Restart blob from RAM. This is mainly useful if you are working on
blob itself because it allows you to download blob and immediately
start it without having to burn it to flash first.


*** "reboot"
------------

This command simply reboots the system.


*** "reload"
------------

Reload blob, kernel, or ramdisk from flash memory to RAM. This command
needs an extra parameter: "blob", "kernel", or "ramdisk". Blob will
respond with:

  blob> reload kernel
  Loading kernel from flash ....... done

The "reload command" will overwrite a just downloaded a kernel or ramdisk.


*** "reset"
-----------

Reset the terminal. This command will write the VT100 reset sequence
(Esc-c) to the terminal. Useful if you forgot to switch your terminal
emulator back to 9600 baud after downloading a kernel or ramdisk.


*** "speed"
-----------

Set the download speed. This command needs a download speed value as
an extra parameter. Valid values are: 1200, 9600, 19200, 38400, 57600,
and 115200. Blob will respond with:

  blob> speed 19200
  Download speed set to 19200 baud


*** "status"
------------

Show the current status. Blob will respond with:

  blob> status
  Bootloader    : blob
  Version       : 2.0.3
  Running from  : internal flash
  Blocksize     : 0x00800000
  Download speed: 115200 baud
  Blob          : from flash
  Kernel        : downloaded, 424333 bytes
  Ramdisk       : from flash

Depending on what commands you used before, these values may or may
not be different.




* Porting blob
==============

Porting blob to a new SA11x0 platform is quite easy and consist of
four steps:

1. Define the features of the architecture
2. Write some architecture specific code
3. Test the new architecture
4. Submit the patch

The next couple of paragraphs describe the process of porting blob to
the "foobar" platform.


** Define the architecture in configure.in
------------------------------------------

First you need two know a couple of things: the name of the board,
what kind of CPU the board uses (SA1100 or SA1110), whether it has
LCD support, and the name of the platform obj and flash obj.
Let's assume the foobar platform has an SA1100 CPU, no use LCD,
and platform obj and flahs obj are foobar.o.
The correct lines for configure.in will be:

    foobar)
	board_name="Foobar Board"
        AC_DEFINE(FOOBAR)
	BLOB_PLATFORM_OBJ="foobar.o"
	BLOB_FLASH_OBJS="foobar.o"
	use_cpu="sa1100"
	use_lcd="no"
        ;;

Put this just after the CreditLART definition.


** Define the architecture in acconfig.h
----------------------------------------

Because configure.in was instructed to define FOOBAR for the foobar
platform, we have to define the symbol in acconfig.h as well. Add the
following two lines to acconfig.h, just after the CLART define:

  /* Define for foobar boards */
  #undef FOOBAR


** Update the build system
--------------------------

Run the following commands to update the configure script,
include/config.h.in, and the Makefile.in files:

  tools/rebuild-gcc
  tools/rebuild-gcc

(yes, twice)


** Configure blob
-----------------

Configure blob for the new foobar architecture:

  setenv CC arm-linux-gcc
  ./configure --with-linux-prefix=/path/to/armlinux/source \
     --with-board=foobar --enable-maintainer-mode \
     --enable-blob-debug arm-unknown-linux-gnu

We're using maintainer-mode and debug information to help the
port. See the section about "Configuring and compiling the package"
for general information.


** Select correct clock speed
-----------------------------

Open src/blob/start.S in an editor, and add a line to select the correct
clock speed (just before the SHANNON definition):

  #if defined FOOBAR
  cpuspeed:       .long   0x09    /* 190 MHz */


** Edit memory settings
-----------------------

Edit src/blob/memsetup-sa1100.S or src/blob/memsetup-sa1110.S,
and add the correct memory setting for the foobar architecture.
Add these (example) settings right before the PLEB definitions:

  #if defined FOOBAR
  mdcas0:         .long   0x1c71c01f
  mdcas1:         .long   0xff1c71c7
  mdcas2:         .long   0xffffffff
  mdcnfg:         .long   0x0334b21f
  #endif

Note that the SA1110 memory settings are not as modular as the SA1100
settings, so you'll have to use your imagination over there to get
proper memory settings.

Right now, the basic blob functionality is ported to your board and
you should be able to compile blob by running "make".


** Edit LED defines
-------------------

If your board has a LED on a GPIO pin, edit include/blob/led.h in an editor
to switch it on early in the boot stage. Let's assume the foobar board
has the LED on GPIO pin 1, so add the following lines just before the
PLEB definition:

  #elif defined FOOBAR
  # define LED_GPIO 0x00000002 /* GPIO 1 */


** Compile blob
---------------

Now compile blob by running:

  make

If everything went right, you have a new blob binary in src/blob.


** Test blob
------------

You are now ready to flash blob to your board and test it. If
something goes wrong in the early boot process, blob will flash the
LED (that's why you should always have a LED on your board), or not
work at all. As soon as you get character on the serial port the most
difficult part is done and you should be ready to port arm-linux to
your board.


** Submit the patch
-------------------

First run "make distclean" in your blob tree so you'll get a clean
source tree. Now rename your current tree and untar the original blob
source (assuming that you're hacking on blob-2.0.3):

  cd ..
  mv blob-2.0.3 blob-2.0.3-foobar
  gzip -dc blob-2.0.3.tar.gz | tar xvf -

Diff the two trees and create a patch file:

  diff -urN blob-2.0.3 blob-2.0.3-foobar > foobar.diff

Now send the patch to me (erikm@users.sourceforge.net) and be sure to
CC a couple of other blob developers (for the current list of blob
developers, see
https://sourceforge.net/project/memberlist.php?group_id=30155 ). The
best way to send the patch is to attach it as plain text to your
message because in that way email clients have less chance to corrupt
the patch.


ChangeLog
(If somebody can show me how to update the ChangeLog from CVS, I'd be
more than grateful -- Erik)

$Id: ChangeLog,v 1.14 2002/01/06 19:03:36 erikm Exp $

blob-2.0.5-pre2:
- MD5 support						Christopher Hoover
- New call command					Christopher Hoover
- Xmodem support					Christopher Hoover
- Serial port driver modularisation			Erik Mouw
- Flash code updates					Christopher Hoover
- HP Labs Badge4 support				Christopher Hoover
- HP Jornada 720 support				Christopher Hoover
- Vercel UD-1 (IDR) support				Tim Riker
- New dump command					Tim Riker
- Fix compile time bug with newer linux kernels		Stefan Eletzhofer
- README update						Kevin Lo

blob-2.0.5-pre1:
- New flash ROM program code				Erik Mouw
- Ipaq H3600 port					Erik Mouw
- Memory tester						Stefan Eletzhofer
- Diagnostics tool (diag)				Stefan Eletzhofer &
							Erik Mouw
- New SA-1110 memory setup code				Stefan Eletzhofer
- PT system 3 port					Stefan Eletzhofer
- Memory test functions					Stefan Eletzhofer
- Enable UART1 if used as the console			Stefan Eletzhofer
- Modular command line parser				Erik Mouw
- Parameter block support				Russ Dill

blob-2.0.4:
- Fix RAMDISK_SIZE (should be in _kilo_ bytes instead
  of bytes)						Erik Mouw
- Add utils directory and populate it with some useful
  scripts						Erik Mouw
- Change asm into asm-arm so the kernel tree doesn't
  need to be configured anymore				Erik Mouw
- Documentation updates					Erik Mouw

blob-2.0.3:
- we moved to SourceForge!
- run from RAM						Erik Mouw & Wookey
- PLEB updates 						Adam Wiggins
- NESA support						Russ Dill
- SHANNON (aka TuxScreen) support 			Tim Riker
- Initial CreditLART support				Erik Mouw
- Resume from suspend					Erik Mouw
- New "reboot" command					Erik Mouw
- Download and flash blob without JTAG dongle		Erik Mouw
- Memory tester						Erik Mouw
- New "reblob" command					Erik Mouw
- Kernel parameter passing				Erik Mouw & Russ Dill

blob-1.0.8-pre2:
- Assabet fixes 					Justin Seger
- RTC fix						Mark Huang

blob-1.0.8-pre1:
- Assabet support					Jeff Sutherland &
							Chester
- Brutus support (without a Brutus)			Erik Mouw
- PLEB support						Erik Mouw

blob-1.07b:
- Boots with linux-2.3.99-pre8-rmk1-np1			Erik Mouw

blob-1.0.6:
- Clock scaling support					Johan Pouwelse

blob-1.0.5:
- Fix flash write bug					Erik Mouw

blob-1.0.0:
- Initial release					Jan-Derk Bakker &
							Erik Mouw


src/blob/main.c
/*-------------------------------------------------------------------------
 * Filename:      main.c
 * Version:       $Id: main.c,v 1.23 2002/01/07 14:48:25 seletz Exp $
 * Copyright:     Copyright (C) 1999, Jan-Derk Bakker
 * Author:        Jan-Derk Bakker <J.D.Bakker@its.tudelft.nl>
 * Description:   Main file for the trivially simple bootloader for the
 *                LART boards
 * Created at:    Mon Aug 23 20:00:00 1999
 * Modified by:   Erik Mouw <J.A.K.Mouw@its.tudelft.nl>
 * Modified at:   Sat Mar 25 14:31:16 2000
 *-----------------------------------------------------------------------*/
/*
 * main.c: main file for the blob bootloader
 * Copyright (C) 2004, 2005 Motorola Inc.
 * Copyright (C) 1999 2000 2001
 *   Jan-Derk Bakker (J.D.Bakker@its.tudelft.nl) and
 *   Erik Mouw (J.A.K.Mouw@its.tudelft.nl)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * 2004-May-18 - (Motorola) Changed enter flash mode
 * 2004-Dec-13 - (Motorola) Added support for command line config and reflash 
                            options Changes made to support our platform
 * 2004-Dec-15 - (Motorola) Barbados power on and hanmdshake 
 * 2005-May-18 - (Motorola) Changes to enable manual flash mode
 *
 */
 

#ident "$Id: main.c,v 1.23 2002/01/07 14:48:25 seletz Exp $"

#ifdef HAVE_CONFIG_H
# include <blob/config.h>
#endif

#include <blob/arch.h>
#include <blob/command.h>
#include <blob/errno.h>
#include <blob/error.h>

#include <blob/init.h>
#include <blob/led.h>
#include <blob/main.h>
#include <blob/md5.h>
#include <blob/md5support.h>
#include <blob/handshake.h>
#include <blob/param_block.h>
#include <blob/serial.h>
#include <blob/time.h>
#include <blob/util.h>
#include <blob/uucodec.h>
#include <blob/xmodem.h>
#include <blob/time.h>
#include <blob/lcd_ezx.h>
#include <blob/bt_uart.h>
#include <linux/string.h>


#include <blob/log.h>
#include <blob/HW_init.h>
#include <blob/SW_init.h>
#include <blob/boot_modes.h>

/* we need update to new config table design */
#define KCT_FLASH_ADDR  	0x67000// new start address of KCT tabe         
#define CMDL_TAG_ADDRESS        KCT_FLASH_ADDR //now, first tag is command line, and its addr is same as kct header       
#define CMDL_TAG_VALUE   	0x8000    
u32 cmd_flag=0;
/* ------------KCT tool functions------------ */
//get KCT head address, possible to a flexible design later
static u32 get_KCT_addr(void){
  return KCT_FLASH_ADDR;

}
//get first tag address based on a offset, possible to a flexible design later
static u32 get_cmdl_tag_addr(u32 offset){
  return CMDL_TAG_ADDRESS;

}
static u16 get_tag_value(u32 tag_addr){
  return *((u16 *)tag_addr);

}
static u16 get_tag_size(u32 tag_addr){
  u16 * p=(u16 *)tag_addr;
  p++;
  return *p;

}
static void* get_tag_content_addr(u32 tag_addr){
  u16 * p=(u16 *)tag_addr;
  p++;				/* skip tag value */
  p++;				/* skip tag size */
  return p;

}
//get and set blob tag in new kct design,
static void parse_blob_tag(void)
{
  u16 tag_val=0, tag_size=0;
  //get kct head addr
  u32 kct_head_addr=get_KCT_addr();

  //get first (command line) tag address
  u32 cmdl_tag_addr = get_cmdl_tag_addr(kct_head_addr);

  //get command_line_value and size
  tag_val=get_tag_value(cmdl_tag_addr);
  logvarhex("parse_blob_tag: tag_val=0x",tag_val);
  tag_size=get_tag_size(cmdl_tag_addr);
  logvarhex("parse_blob_tag: tag_size=0x",tag_size);
  //set first tag if possible

  if (CMDL_TAG_VALUE ==tag_val){
    char * cmdline=get_tag_content_addr(cmdl_tag_addr);
    cmd_flag = 1;
    memcpy(blob_status.cmdline, cmdline, tag_size);
    logstr("parse_blob_tag:cmdl ok\n");
  }else {
    cmd_flag = 0;
    blob_status.cmdline[0] = '\0';
  }

  //go to next tag....

}

/* end of -------KCT tool functions----------- */
#define FLASH_TOOL_TEST_VERSION
#define BARBADOS_HANDSHAKE
#ifdef BARBADOS_HANDSHAKE  //waiting for about 4s
#define BARBADOS_BP_RDY_TIMEOUT  40000000
#endif
//#define OUTPUT_LOG_TO_STPORT

#ifdef MBM
extern int boot_linux(int,char*);

#endif

#ifdef FLASH_TOOL_TEST_VERSION

char value_trans_to_char(u8 data )
{
	char ret ='F';

        if(data>15)
		return ret;

        if( data <10) {
		ret = '0'+data;
        } else {
		ret = 'a'+data-10;
        }

        return ret;
}

int translate_xnum_to_string( u8 * pStr,u32 num)
{
        int i;

        pStr[8]=0;
        for(i=0;i<8;i++) {
		pStr[7-i] = value_trans_to_char(num%16);
		num = num/16;           
        }
        return 0;
}

/* output reset information for debug */
void output_reset_inf(u32 sleepflag)
{
        u8 mystr[20];

        printlcd("\nFLAG=0x");
        translate_xnum_to_string( mystr,sleepflag);  
        printlcd(mystr);

        printlcd("\nRESET=0x");
        translate_xnum_to_string( mystr,FFSPR);  
        printlcd(mystr);

        printlcd("\nBRESET=0x");        
        translate_xnum_to_string( mystr,PSPR);  
        printlcd(mystr);

        printlcd("\nARESET=0x");        
        translate_xnum_to_string( mystr,RCSR);  
        printlcd(mystr);
}

/* save reset cause at previous time for debug */
void save_reset_inf(void)
{
        PSPR = RCSR;
}


#else

#define output_reset_inf(sleepflag)
#define save_reset_inf()
	
#endif


blob_status_t blob_status;

static int do_reload(char *what);

static char *boot_params[] = { // ALEX
	"root=3e00", 
	"rootfstype=cramfs",
	"ip=off",
	"paniclog=on",
};

#define NR_BOOT_PARAMS 4 //(sizeof(boot_params) / sizeof(char *))

int main(void)
{
	int retval = 0;

    u32 sleepflag;
#ifdef PARAM_START
	u32 conf;
#endif

	/* do hw init for product */
//	hw_init(0);  // ALEX don't need to initial hardware again

	/* do sw init for product */
//	sw_init(0);
	/* parse the core tag, for critical things like terminal speed */

#if 1 // ALEX test codes
{
	enter_simple_pass_through_mode();
}
#else
EnableLCD_8bit_active();
#endif // ALEX

#if 0
#ifdef PARAM_START
	parse_ptag((void *) PARAM_START, &conf);
#endif	
#if 1 //ALEX
  	printlcd("parse_blob_tag\n");      
	parse_blob_tag();
#endif

        sleepflag = *(unsigned long *)(FLAG_ADDR);
#ifndef HAINAN
	if ( check_flash_flag() )
	  enter_tcmd_flash_mode(0);
#endif	
  	printlcd("check_pass_through_flag\n");      
	/* add decision to pass-through mode */
	if(check_pass_through_flag())
	  enter_pass_through_mode();

#ifndef HAINAN
        if ( !check_valid_code()|| check_manual_flash_mode() ) 
	  enter_manual_flash_mode(0);
#endif
#endif
  	printlcd("loading kernel\n");      
#if 1 //ALEX
	/* Load kernel and ramdisk from flash to RAM */
	do_reload("kernel");
//      logstr("\nezx_test: Now, boot linux!\n");
      printlcd("ezx_test: Now, boot linux!\n");
//void (*theKernel)(void) = KERNEL_RAM_BASE;
//  	theKernel();
#endif
{
	long *p = (long *)0xa0800;
	int i;
	for (i = 0; i < 1000000; i++) {
		printf("%08x, %08x\n", *p++, *p++); 
		Delay(50);
	}

	while (1);
}

	save_reset_inf();

      printlcd("boot\n");
	if(retval == 0) {
#ifndef MBM
	  parse_command("boot");
#else
	  boot_linux(0,0);
//	  boot_linux(NR_BOOT_PARAMS+1, boot_params);
#endif
	}
printlcd("what????\n");
while(1);
	return 0;
} /* main */



static int do_reload(char *what)
{
	u32 *dst = 0;
	u32 *src = 0;
	int numWords;

	if(strncmp(what, "blob", 5) == 0) {
		dst = (u32 *)BLOB_RAM_BASE;
		src = (u32 *)BLOB_FLASH_BASE;
		numWords = BLOB_FLASH_LEN / 4;
		blob_status.blobSize = 0;
		blob_status.blobType = fromFlash;
		//SerialOutputString("Loading blob from flash ");
#ifdef PARAM_START
	} else if(strncmp(what, "param", 6) == 0) {
		dst = (u32 *)PARAM_RAM_BASE;
		src = (u32 *)PARAM_FLASH_BASE;
		numWords = PARAM_FLASH_LEN / 4;
		blob_status.paramSize = 0;
		blob_status.paramType = fromFlash;
		//SerialOutputString("Loading paramater block from flash ");
#endif
	} else if(strncmp(what, "kernel", 7) == 0) {
		dst = (u32 *)KERNEL_RAM_BASE;
		src = (u32 *)KERNEL_FLASH_BASE;
		numWords = KERNEL_FLASH_LEN / 4;
		blob_status.kernelSize = 0;
		blob_status.kernelType = fromFlash;
		//SerialOutputString("Loading kernel from flash ");
	} else if(strncmp(what, "ramdisk", 8) == 0) {
		dst = (u32 *)RAMDISK_RAM_BASE;
		src = (u32 *)RAMDISK_FLASH_BASE;
		numWords = RAMDISK_FLASH_LEN / 4;
		blob_status.ramdiskSize = 0;
		blob_status.ramdiskType = fromFlash;
		//SerialOutputString("Loading ramdisk from flash ");
	} else {
		printerror(EINVAL, what);
		return 0;
	}

	MyMemCpy(dst, src, numWords);
	//SerialOutputString(" done\n");

	return 0;
}


	

